extensions/attributegen/plugin.cc

```diff
@@ -0,0 +1,147 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "extensions/attributegen/plugin.h"
+
+// WASM_PROLOG
+#ifndef NULL_PLUGIN
+
+#else  // NULL_PLUGIN
+
+#include "extensions/common/wasm/null/null.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+namespace Null {
+namespace Plugin {
+
+#endif  // NULL_PLUGIN
+
+// END WASM_PROLOG
+
+namespace AttributeGen {
+
+bool PluginRootContext::onConfigure(size_t) {
+  std::unique_ptr<WasmData> configuration = getConfiguration();
+  // Parse configuration JSON string.
+  JsonParseOptions json_options;
+  json_options.ignore_unknown_fields = true;
+  istio::attributegen::PluginConfig config;
+  Status status =
+      JsonStringToMessage(configuration->toString(), &config, json_options);
+  if (status != Status::OK) {
+    LOG_WARN(absl::StrCat(
+        "Cannot parse plugin configuration JSON string [YAML not supported]: ",
+        configuration->toString()));
+    return false;
+  }
+
+  return initAttributeGen(config);
+}
+
+bool PluginRootContext::initAttributeGen(
+    const istio::attributegen::PluginConfig& config) {
+  // Clear prior host resources.
+  // TODO(mjog) see if this still needed since onDone is always called.
+  cleanupAttributeGen();
+
+  for (auto agconfig : config.attributes()) {
+    EvalPhase phase = on_log;
+    if (agconfig.phase() == istio::attributegen::ON_REQUEST) {
+      phase = on_request;
+    }
+    auto ag = AttributeGenerator(phase, agconfig.output_attribute());
+    for (auto matchconfig : agconfig.match()) {
+      uint32_t token = 0;
+      if (matchconfig.condition().empty()) {
+        ag.add_match(Match("", 0, matchconfig.value()));
+        continue;
+      }
+      if (createExpression(matchconfig.condition(), &token) != WasmResult::Ok) {
+        LOG_WARN(absl::StrCat("Cannot create expression: <",
+                              matchconfig.condition(), "> for ",
+                              agconfig.output_attribute()));
+        return false;
+      }
+      LOG_DEBUG(absl::StrCat("Added ", agconfig.output_attribute(), " if (",
+                             matchconfig.condition(), ") -> ",
+                             matchconfig.value()));
+      ag.add_match(Match(matchconfig.condition(), token, matchconfig.value()));
+    }
+    gen_.push_back(ag);
+  }
+  return true;
+}
+
+void PluginRootContext::cleanupAttributeGen() {
+  for (auto ag : gen_) {
+    ag.cleanup();
+  }
+  gen_.clear();
+}
+
+// attributeGen is called on the data path.
+void PluginRootContext::attributeGen(EvalPhase phase) {
+  for (const auto& ag : gen_) {
+    if (phase != ag.phase()) {
+      continue;
+    }
+
+    std::string val;
+    auto eval_status = ag.evaluate(&val);
+    if (!eval_status) {
+      LOG_DEBUG(absl::StrCat("Failed eval. setting ", ag.error_attribute()));
+      // eval failed set error attribute
+      setFilterState(ag.error_attribute(), "1");
```

Comment:
Good point about `_error` attribute overshadowing an attribute that is already there.
However since this is part of the contract and API, use will hopefully see this.

We always *need* to filterstate or error so that downstream plugins can distinguish between an error and no-match.

Reply:
It's unclear why we need error filter state. Filter state is expensive, and just putting "1" there is wasteful.

Reply:
Please suggest something else that will help the consumer of this attribute distinguish between attribute not set because condition did not match, and attribute could not be evaluated.
The consuming plugin may be authz and it is important to distinguish between the two.

Reply:
If it's not set, then it's not set. The policy would just evaluate to an error, or use presence the checks for extra control in rare cases.

Reply:
it is no more wasteful than setting the attribute itself.